environment.py

```diff
@@ -64,14 +64,19 @@ def env() -> Mapping[str, Optional[str]]:
         # TDR source = 'tdr:', Google Cloud project name,
         #              ':', ( 'dataset' | 'snapshot' ),
         #              '/', TDR dataset or snapshot name,
-        #              ':', [ UUID prefix [ '/', Partition prefix length ] ] ;
+        #              ':', [ UUID prefix [ '/', partition prefix length ] ] ;
         #
         # canned source = 'https://github.com',
         #                 '/', owner,
         #                 '/', repo,
         #                 '/tree/', ref,
-        #                 ['/', path] ;
-        #                 ':', [ UUID prefix [ '/', Partition prefix length ] ] ;
+        #                 ['/', path],
+        #                 ':', [ UUID prefix [ '/', partition prefix length ] ] ;
+        #
+        # digit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9" ;
+        # prefix character = digit | "A" | "B" | "C" | "D" | "E" | "F" ;
+        # UUID prefix = prefix character, 7 * [ prefix character ] ;
+        # partition prefix length = digit - "9" ;
```

Comment:
I am not sure if we should even go to this amount of detail since it still doesn't correctly express all the constraints. For example, the grammar treats `fffffff/8` as valid but the code would reject it. I think you should describe the prefix and length informally.